File: sdk/wat_sdk.js
'use strict';

var fs          = require('fs');
var crypto      = require('crypto');
var http        = require('http');
var https       = require('https');
var querystring = require('querystring');

var UPLOAD_DISABLED = false;
try {
 var FormData = require('form-data');
} catch(err) {
  UPLOAD_DISABLED = true;
  console.log('Cannot import `FormData` module, the WATClient.upload() method is disabled.')
}

var COLOR_MAP = {
  'grey'   : '\x1b[0;30m',
  'red'    : '\x1b[0;31m',
  'green'  : '\x1b[0;32m',
  'yellow' : '\x1b[0;33m',
  'blue'   : '\x1b[0;34m',
  'magenta': '\x1b[0;35m',
  'cyan'   : '\x1b[0;36m',
}

function strf() {
  var args = Array.prototype.slice.call(arguments);
  if (0 === args.length) {
    return '';
  }

  var pattern = args.shift();
  try {
    pattern = pattern.toString();
  } catch (err) {
    pattern = '';
  }

  return pattern.replace(/\{(\d+)\}/g, function replaceFunc(m, i) {
    return args[i] + '';
  });
};

function colored(s, color) {
  if (!color) {
    color = 'yellow';
  }

  color = COLOR_MAP[color];

  return color + s + '\x1b[0m';
};

var WATClient = function(options) {
  this.debug = options.debug || false;

  this.akId     = options.akId;
  this.akSecret = options.akSecret;

  this.host = options.host;
  this.port = options.port;

  if (this.host && this.host.indexOf(':') > -1) {
    var host_port = this.host.split(':');
    this.host = host_port[0];
    this.port = parseInt(host_port[1]);
  }

  this.timeout  = options.timeout  || 3;
  this.useHTTPS = options.useHTTPS || false;

  if (!this.port) {
    if (this.useHTTPS) {
      this.port = 443;
    } else {
      this.port = 80;
    }
  }

  this.headerFields = {
      akId       : 'X-Wat-Ak-Id',
      akTimestamp: 'X-Wat-Ak-Timestamp',
      akNonce    : 'X-Wat-Ak-Nonce',
      akSign     : 'X-Wat-Ak-Sign',
      traceId    : 'X-Trace-Id',
  }
  if (options.headerFields) {
    Object.assign(this.headerFields, options.headerFields);
  }
};

WATClient.prototype.getSign = function(method, path, timestamp, nonce) {
  if (!timestamp) {
    timestamp = parseInt(Date.now() / 1000).toString();
  }

  if (!nonce) {
    nonce = Math.random().toString();
  }

  var stringToSign = [ timestamp, nonce, method.toUpperCase(), path ].join('&');

  if (this.debug) {
    console.log(strf('{0} {1}',
      colored('[String to Sign]'),
      stringToSign));
  }

  var c = crypto.createHmac('sha1', this.akSecret);
  c.update(stringToSign);
  var sign = c.digest('hex');

  return sign;
};

WATClient.prototype.verifySign = function(sign, method, path, timestamp, nonce) {
  var expectedSign = this.getSign(method, path, timestamp, nonce);

  return (sign === expectedSign);
};

WATClient.prototype.getAuthHeader = function(method, path) {
  var timestamp = parseInt(Date.now() / 1000).toString();
  var nonce     = Math.random().toString();

  var sign = this.getSign(method, path, timestamp, nonce);

  var authHeader = {}
  authHeader[this.headerFields.akId]        = this.akId;
  authHeader[this.headerFields.akTimestamp] = timestamp;
  authHeader[this.headerFields.akNonce]     = nonce;
  authHeader[this.headerFields.akSign]      = sign;

  return authHeader;
};

WATClient.prototype.verifyAuthHeader = function(headers, method, path) {
  var timestamp = headers[this.headerFields.akTimestamp.toLowerCase()] || '';
  var nonce     = headers[this.headerFields.akNonce.toLowerCase()]     || '';
  var sign      = headers[this.headerFields.akSign.toLowerCase()]      || '';

  return this.verifySign(sign, method, path, timestamp, nonce);
};

WATClient.prototype.run = function(options, callback) {
  var method  = options.method || 'GET';
  var path    = options.path;
  var query   = options.query;
  var body    = options.body;
  var headers = options.headers;
  var traceId = options.traceId;

  // Prepare method/query/body
  method = method.toUpperCase();

  if (query) {
    path = path + '?' + querystring.stringify(query);
  }

  if (body) {
    if ('object' === typeof body) {
      body = JSON.stringify(body);
    } else {
      body = '' + (body || '');
    }
  }

  if (this.debug) {
    console.log('='.repeat(50));
    console.log(strf('{0} {1} {2}', colored('[Request]'), colored(method, 'cyan'), colored(path, 'cyan')));
    if (body) {
      console.log(strf('{0} {1}', colored('[Payload]'), body));
    }
  }

  // Prepare headers with auth info
  headers = headers || {};
  headers['Content-Type'] = 'application/json';
  if (traceId) {
    headers[this.headerFields.traceId] = traceId;
  }

  if (this.akId && this.akSecret) {
    var authHeader = this.getAuthHeader(method, path);
    Object.assign(headers, authHeader);
  }

  // Do HTTP/HTTPS
  var requestOptions = {
    host   : this.host,
    port   : this.port,
    path   : path,
    method : method,
    headers: headers,
    timeout: this.timeout,
  };

  var httpLib = this.useHTTPS ? https : http;

  var self = this;

  var respStatusCode = 0;
  var respRawData    = '';
  var respData       = '';
  var req = httpLib.request(requestOptions, function(res) {
    respStatusCode = res.statusCode;

    res.on('data', function(chunk) {
      respRawData += chunk;
    });

    res.on('end', function() {
      var respContentType = res.headers['content-type'].split(';')[0].trim();
      if (respContentType === 'application/json') {
        respData = JSON.parse(respRawData);
      }

      if (self.debug) {
        var _color = respStatusCode >= 400 ? 'red' : 'green';

        console.log(colored(strf('{0} {1}', '[Response]', respStatusCode), _color));
        console.log(colored(strf('{0} {1}', '[Payload]', respRawData.toString()), _color));
      }

      if ('function' === typeof callback) {
        return callback(null, respData, respStatusCode);
      }
    });
  });

  req.on('error', function(err) {
    if ('function' === typeof callback) {
      return callback(err);
    }
  });

  if (body) {
    req.write(body);
  }

  req.end();
};

WATClient.prototype.upload = function(options, callback) {
  if (UPLOAD_DISABLED) {
    throw Error('`WATClient.upload()` method need `FormData` module.')
  }

  var path        = options.path;
  var fileBuffer  = options.fileBuffer;
  var filename    = options.filename;
  var query       = options.query;
  var fields      = options.fields;
  var headers     = options.headers;
  var traceId     = options.traceId;

  // Prepare method/query/fields/file
  var method = 'POST';

  if (!filename) {
    filename = 'uploadfile';
  }
  filename = querystring.escape(filename);

  if (query) {
    path = path + '?' + querystring.stringify(query);
  }

  if (fields && 'object' !== typeof fields) {
    throw Error('`fields` should be a plain JSON')
  }

  // FormData
  var formData = new FormData();

  if (fields) {
    for (var k in fields) if (fields.hasOwnProperty(k)) {
      formData.append(k, fields[k]);
    }
  }

  formData.append('files', fileBuffer, {filename: filename});

  if (this.debug) {
    console.log('='.repeat(50));
    console.log(strf('{0} {1} {2}', colored('[Request]'), colored(method, 'cyan'), colored(path, 'cyan')));
    if (fields) {
      console.log(strf('{0} {1}', colored('[Fields]'), fields));
    }
    if (fileBuffer) {
      console.log(strf('{0} {1}', colored('[File]'), filename));
    }
  }

  // Prepare headers with auth info
  headers = headers || {};
  headers['Content-Type'] = 'application/json';
  if (traceId) {
    headers[this.headerFields.traceId] = traceId;
  }

  if (this.akId && this.akSecret) {
    var authHeader = this.getAuthHeader(method, path);
    Object.assign(headers, authHeader);
  }

  var formHeader = formData.getHeaders();
  Object.assign(headers, formHeader);

  // Do HTTP/HTTPS
  var requestOptions = {
    host   : this.host,
    port   : this.port,
    path   : path,
    method : method,
    headers: headers,
    timeout: this.timeout,
  };

  var httpLib = this.useHTTPS ? https : http;

  var self = this;

  var respStatusCode = 0;
  var respRawData    = '';
  var respData       = '';
  var req = httpLib.request(requestOptions, function(res) {
    respStatusCode = res.statusCode;

    res.on('data', function(chunk) {
      respRawData += chunk;
    });

    res.on('end', function() {
      var respContentType = res.headers['content-type'].split(';')[0].trim();
      if (respContentType === 'application/json') {
        respData = JSON.parse(respRawData);
      }

      if (self.debug) {
        var _color = respStatusCode >= 400 ? 'red' : 'green';

        console.log(colored(strf('{0} {1}', '[Response]', respStatusCode), _color));
        console.log(colored(strf('{0} {1}', '[Payload]', respRawData.toString()), _color));
      }

      if ('function' === typeof callback) {
        return callback(null, respData, respStatusCode);
      }
    });
  });

  req.on('error', function(err) {
    if ('function' === typeof callback) {
      return callback(err);
    }
  });

  if (formData.getLengthSync() > 0) {
    formData.pipe(req);
  }

  req.end();
};

WATClient.prototype.get = function(options, callback) {
  options = options || {};
  options.method = 'GET';
  return this.run(options, callback);
};

WATClient.prototype.post = function(options, callback) {
  options = options || {};
  options.method = 'POST';
  return this.run(options, callback);
};

WATClient.prototype.put = function(options, callback) {
  options = options || {};
  options.method = 'PUT';
  return this.run(options, callback);
};

WATClient.prototype.delete = function(options, callback) {
  options = options || {};
  options.method = 'DELETE';
  return this.run(options, callback);
};

exports.WATClient = WATClient;

if (require.main === module) {
  function testSuit(name, callback) {
    var clientOpt = {
      akId    : 'ak-7Qf3KXH8QZOrW8Tf',
      akSecret: 'WaYGi4cBsievlfZsNhE3fY40ZB9dI9L3',
      host    : 'ubuntu18-dev.vm',
      port    : 80,
    };
    var c = new WATClient(clientOpt);
    c.debug = true;

    // 1. GET Request Test
    var opt1 = {
      path   : '/api/v1/do/ping',
      traceId: 'TEST-NODE-001',
    };
    c.get(opt1, function(err, respData, respStatusCode) {

      // 2. POST Request Test
      var opt2 = {
        path: '/api/v1/do/echo',
        body: {
          'echo': {
            'int'    : 1,
            'str'    : 'Hello World',
            'unicode': '你好，世界！',
            'none'   : null,
            'boolean': true,
          }
        },
        traceId: 'TEST-NODE-002',
      };
      c.post(opt2, function(err, respData, respStatusCode) {

        // 3. UPLOAD Request Test
        if (UPLOAD_DISABLED) return callback && callback();

        var opt3 = {
          path      : '/api/v1/files/do/upload',
          fileBuffer: fs.readFileSync('使用说明_SDK.md'),
          filename  : '使用说明_SDK.md',
          fields    : {
            'note': 'This is a Chinese README. 这是一份中文使用说明。'
          },
          traceId: 'TEST-NODE-003',
        };
        c.upload(opt3, function(err, respData, respStatusCode) {

          return callback && callback();
        });
      });
    });
  };

  testSuit('test', function() {
    testSuit(null);
  });
};

File: sdk/wat_sdk.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import time
import uuid
import json
import hmac
from hashlib import sha1, md5
import types
import base64

try:
    import httplib
    from urllib import urlencode, quote
    from urlparse import urlparse
except ImportError:
    import http.client as httplib
    from urllib.parse import urlparse, urlencode, quote

UPLOAD_DISABLED = False
try:
    import requests
except ImportError:
    UPLOAD_DISABLED = True
    print('Cannot import `requests` module, the WATClient.upload() method is disabled.')

COLOR_MAP = {
    'grey'   : '\033[0;30m',
    'red'    : '\033[0;31m',
    'green'  : '\033[0;32m',
    'yellow' : '\033[0;33m',
    'blue'   : '\033[0;34m',
    'magenta': '\033[0;35m',
    'cyan'   : '\033[0;36m',
}

# -----------------------------------------------
# Python2 ~ Python3 Compatibility Code From `six`
# -----------------------------------------------
#
# Copyright (c) 2010-2019 Benjamin Peterson
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

PY2 = sys.version_info[0] == 2
PY3 = sys.version_info[0] == 3
PY34 = sys.version_info[0:2] >= (3, 4)

if PY3:
    string_types = str,
    integer_types = int,
    class_types = type,
    text_type = str
    binary_type = bytes
else:
    string_types = basestring,
    integer_types = (int, long)
    class_types = (type, types.ClassType)
    text_type = unicode
    binary_type = str

def ensure_binary(s, encoding='utf-8', errors='strict'):
    if isinstance(s, text_type):
        return s.encode(encoding, errors)
    elif isinstance(s, binary_type):
        return s
    else:
        raise TypeError("not expecting type '%s'" % type(s))

def ensure_str(s, encoding='utf-8', errors='strict'):
    if not isinstance(s, (text_type, binary_type)):
        raise TypeError("not expecting type '%s'" % type(s))
    if PY2 and isinstance(s, text_type):
        s = s.encode(encoding, errors)
    elif PY3 and isinstance(s, binary_type):
        s = s.decode(encoding, errors)
    return s

def colored(s, color=None):
    if not color:
        color = 'yellow'

    color = COLOR_MAP[color]

    return color + '{}\033[0m'.format(s)

class WATClient(object):
    def __init__(self, ak_id=None, ak_secret=None, host=None, port=None, timeout=3, use_https=False, header_fields=None, debug=False):
        self.debug = debug

        self.ak_id     = ak_id
        self.ak_secret = ak_secret

        self.host = host
        self.port = port

        if self.host and ':' in self.host:
            host_port = self.host.split(':')
            self.host = host_port[0]
            self.port = int(host_port[1])

        self.timeout   = timeout
        self.use_https = use_https

        if not self.port:
            if self.use_https:
                self.port = 443
            else:
                self.port = 80

        self.header_fields = {
            'akId'       : 'X-Wat-Ak-Id',
            'akTimestamp': 'X-Wat-Ak-Timestamp',
            'akNonce'    : 'X-Wat-Ak-Nonce',
            'akSign'     : 'X-Wat-Ak-Sign',
            'traceId'    : 'X-Trace-Id',
        }
        if header_fields:
            self.header_fields.update(header_fields)

    def get_sign(self, method, path, timestamp, nonce):
        if timestamp is None:
            timestamp = str(int(time.time()))

        if nonce is None:
            nonce = str(uuid.uuid4().hex)

        string_to_sign = '&'.join([ timestamp, nonce, method.upper(), path])

        if self.debug:
            print('{} {} {}'.format(
                colored('[String to Sign]'),
                colored(self.ak_sign_version, 'cyan'),
                ensure_str(string_to_sign)))

        ak_secret      = ensure_binary(self.ak_secret)
        string_to_sign = ensure_binary(string_to_sign)

        h = hmac.new(ak_secret, string_to_sign, sha1)

        sign = h.hexdigest()
        sign = ensure_str(sign)

        return sign

    def verify_sign(self, sign, method, path, timestamp, nonce):
        expected_sign = self.get_sign(method, path, timestamp, nonce)

        return (sign == expected_sign)

    def get_auth_header(self, method, path):
        timestamp = str(int(time.time()))
        nonce     = uuid.uuid4().hex

        sign = self.get_sign(method, path, timestamp, nonce)

        auth_headers = {
            self.header_fields['akId']       : self.ak_id,
            self.header_fields['akTimestamp']: timestamp,
            self.header_fields['akNonce']    : nonce,
            self.header_fields['akSign']     : sign,
        }
        return auth_headers

    def verify_auth_header(self, headers, method, path):
        timestamp = headers.get(self.header_fields['akTimestamp'], '')
        nonce     = headers.get(self.header_fields['akNonce'], '')
        sign      = headers.get(self.header_fields['akSign'], '')

        return self.verify_sign(sign, method, path, timestamp, nonce)

    def run(self, method, path, query=None, body=None, headers=None, trace_id=None):
        # Prepare method/query/body
        method = method.upper()

        if query:
            path = path + '?' + urlencode(query)

        if body:
            if isinstance(body, (tuple, list, dict)):
                body = json.dumps(body, sort_keys=True, separators=(',', ':'))

            body = ensure_binary(body)

        if self.debug == True:
            print('=' * 50)
            print('{} {} {}'.format(colored('[Request]'), colored(method, 'cyan'), colored(ensure_str(path), 'cyan')))
            if body:
                print('{} {}'.format(colored('[Payload]'), ensure_str(body)))

        # Prepare headers with auth info
        headers = headers or {}
        headers['Content-Type'] = 'application/json'
        if trace_id:
            headers[self.header_fields['traceId']] = trace_id

        if self.ak_id and self.ak_secret:
            auth_headers = self.get_auth_header(method, path)
            headers.update(auth_headers)

        # Do HTTP/HTTPS
        conn = None
        if self.use_https:
            conn = httplib.HTTPSConnection(self.host, port=self.port, timeout=self.timeout)
        else:
            conn = httplib.HTTPConnection(self.host, port=self.port, timeout=self.timeout)

        conn.request(method, path, body=body, headers=headers)

        # Get response
        resp = conn.getresponse()

        resp_status_code = resp.status
        resp_raw_data    = resp.read()

        resp_content_type = resp.getheader('Content-Type')
        if isinstance(resp_content_type, string_types):
            resp_content_type = resp_content_type.split(';')[0].strip()

        if resp_content_type == 'application/json':
            resp_data = json.loads(ensure_str(resp_raw_data))
        else:
            resp_data = resp_raw_data

        if self.debug == True:
            _color = 'green'
            if resp_status_code >= 400:
                _color = 'red'

            print(colored('{} {}'.format('[Response]', resp_status_code), _color))
            print(colored('{} {}'.format('[Payload]', ensure_str(resp_raw_data)), _color))

        return resp_status_code, resp_data

    def upload(self, path, file_buffer, filename=None, query=None, fields=None, headers=None, trace_id=None):
        if UPLOAD_DISABLED:
            raise Exception('`WATClient.upload()` method need `requests` module.')

        # Prepare method/query/fields/file
        method = 'POST'

        if not filename:
            filename = 'uploadfile'

        filename = quote(filename)

        if query:
            path = path + '?' + urlencode(query)

        if fields and not isinstance(fields, dict):
            raise Exception('`fields` should be a plain JSON')

        files = {'files': (filename, file_buffer)}

        if self.debug == True:
            print('=' * 50)
            print('{} {} {}'.format(colored('[Request]'), colored(method, 'cyan'), colored(ensure_str(path), 'cyan')))
            if fields:
                print('{} {}'.format(colored('[Fields]'), ensure_str(json.dumps(fields))))
            if files:
                print('{} {}'.format(colored('[File]'), filename))

        # Preapre HTTP
        url = 'http://{}:{}{}'.format(self.host, self.port, path)
        req = requests.Request('POST', url, data=fields, files=files)
        prepared_req = req.prepare()

        # Prepare headers with auth info
        headers = headers or {}
        if trace_id:
            headers[self.header_fields['traceId']] = trace_id

        if self.ak_id and self.ak_secret:
            auth_headers = self.get_auth_header(method, path)
            headers.update(auth_headers)

        prepared_req.headers.update(headers)

        with requests.Session() as s:
            resp = s.send(prepared_req)

        # Get response
        resp_status_code = resp.status_code
        resp_raw_data    = resp.text

        resp_content_type = resp.headers.get('Content-Type', '').split(';')[0].strip()
        if resp_content_type == 'application/json':
            resp_data = json.loads(resp_raw_data)

        if self.debug == True:
            _color = 'green'
            if resp_status_code >= 400:
                _color = 'red'

            print(colored('{} {}'.format('[Response]', resp_status_code), _color))
            print(colored('{} {}'.format('[Payload]', ensure_str(resp_raw_data)), _color))

        return resp_status_code, resp_data

    def get(self, path, query=None, headers=None, trace_id=None):
        return self.run('GET', path, query, None, headers, trace_id=trace_id)

    def post(self, path, query=None, body=None, headers=None, trace_id=None):
        return self.run('POST', path, query, body, headers, trace_id=trace_id)

    def put(self, path, query=None, body=None, headers=None, trace_id=None):
        return self.run('PUT', path, query, body, headers, trace_id=trace_id)

    def delete(self, path, query=None, headers=None, trace_id=None):
        return self.run('DELETE', path, query, None, headers, trace_id=trace_id)

if __name__ == '__main__':
    c = WATClient('ak-7Qf3KXH8QZOrW8Tf', 'WaYGi4cBsievlfZsNhE3fY40ZB9dI9L3', host='ubuntu18-dev.vm:80')
    c.debug = True

    # 1. GET Request Test
    status_code, resp = c.get('/api/v1/do/ping',
            trace_id='TEST-PY{}-001'.format(sys.version_info[0]))

    # 2. POST Request Test
    body = {
        'echo': {
            'int'    : 1,
            'str'    : 'Hello World',
            'unicode': u'你好，世界！',
            'none'   : None,
            'boolean': True,
        }
    }

    status_code, resp = c.post('/api/v1/do/echo',
            body=body,
            trace_id='TEST-PY{}-002'.format(sys.version_info[0]))

    # 3. UPLOAD Request Test
    if not UPLOAD_DISABLED:
        with open('使用说明_SDK.md', 'rb') as _f:
            file_buffer = _f.read()

        fields = {
            'note': 'This is a Chinese README. 这是一份中文使用说明。'
        }

        status_code, resp = c.upload('/api/v1/files/do/upload',
            file_buffer=file_buffer, filename='使用说明_SDK.md', fields=fields,
            trace_id='TEST-PY{}-003'.format(sys.version_info[0]))
